### DISCUSSION TITLE
[Feature]: Http2 proxy suport

### DISCUSSION THREAD

**Discussion by winchesHe** [web-infra-dev/rsbuild#4124] [General]
### What problem does this feature solve?

When use the https and proxy it will auto change to use http/1

### What does the proposed API look like?

1. It can support current proxy config
2. One more `attr` in `serverConfig` like `http2: boolean` which can decide if the http2 open
3. Want to use [http2-proxy](https://github.com/nxtedition/node-http2-proxy) to create a proxy middleware

**Comment by winchesHe** [comment]
If there is such a requirement, I am willing to implement it

**Comment by chenjiahan** [comment]
[node-http2-proxy] is not under active development, I'm not sure if adopting it is a good idea.

There are some related discussion:

- https://github.com/chimurai/http-proxy-middleware/issues/214
- https://github.com/vitejs/vite/issues/4184
- https://github.com/chimurai/http-proxy-middleware/issues/426

**Comment by MirKml** [comment]
hey there is a huge step forward - new modern proxy library was created - [http-proxy-3](https://www.npmjs.com/package/http-proxy-3) and it is used newly in vite https://github.com/vitejs/vite/pull/20402 
http/2 support included
it seems that use http-proxy-3 is way to go

**Comment by A933997** [comment]
> ### What problem does this feature solve?
> 
> When use the https and proxy it will auto change to use http/1
> 
> ### What does the proposed API look like?
> 
> 1. It can support current proxy config
> 2. One more `attr` in `serverConfig` like `http2: boolean` which can decide if the http2 open
> 3. Want to use [http2-proxy](https://github.com/nxtedition/node-http2-proxy) to create a proxy middleware

**Comment by chenjiahan** [comment]
This depends on https://github.com/chimurai/http-proxy-middleware/pull/1159 or https://github.com/chimurai/http-proxy-middleware/pull/1160 being merged first.
